feat: ability to add people to team (and org) via round robin

Amit91848 opened the pull request:
## What does this PR do?

Fixes #14169
Fixes #13532 

For normal users

https://github.com/calcom/cal.com/assets/74371312/999c1eb1-16a4-47cb-a8b7-a776d04fe91f

For orgs

https://github.com/calcom/cal.com/assets/74371312/53466bad-ad86-4de8-b2cf-f884293c7b26




## Type of change

- New feature (non-breaking change which adds functionality)

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

Amit91848 committed:
feat: ability to add people to team (and org) via round robin

SomayChauhan commented:
hi @Amit91848 
check out this [LOOM](https://www.loom.com/share/5488f1d1b5b741f59ad73899856252d3?sid=f2265ef6-1b22-4326-866b-7cf7e213ea85)
should show the invited member in the dropdown

Amit91848 commented:
> hi @Amit91848 check out this [LOOM](https://www.loom.com/share/5488f1d1b5b741f59ad73899856252d3?sid=f2265ef6-1b22-4326-866b-7cf7e213ea85) should show the invited member in the dropdown

Won't the users have to accept the invite first to be shown and assigned events?

Amit91848 commented:
Will do

Amit91848 committed:
fix: show invited pending members

Amit91848 commented:
@SomayChauhan Fixed both, you can now see invited members and also cannot invite them again


https://github.com/calcom/cal.com/assets/74371312/acbb919a-dd35-4413-b376-16c81ea53a5e

SomayChauhan commented:
![Screenshot 2024-03-23 at 12 48 48 PM](https://github.com/calcom/cal.com/assets/32706411/fb811f7e-f357-4821-b4c5-5e5a43209ab6)
a pending badge like this will look better i think
also i should be able to select the pending members.
![Screenshot 2024-03-23 at 12 51 05 PM](https://github.com/calcom/cal.com/assets/32706411/c538241e-98c8-4dde-ae79-2c32972fb5d3)
should show the pending badge here as well for pending members.

SomayChauhan commented:
additionally this would be Nice to have, adding comma seperated emails opens the bulk emails by default
[LOOM](https://www.loom.com/share/9acd2ce6dbf141d89af4b9125724aa74?sid=018c0b96-d3c5-447f-a56e-0780037bd180)

Amit91848 commented:
> ![Screenshot 2024-03-23 at 12 48 48 PM](https://private-user-images.githubusercontent.com/32706411/316224974-fb811f7e-f357-4821-b4c5-5e5a43209ab6.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExNzkwMzcsIm5iZiI6MTcxMTE3ODczNywicGF0aCI6Ii8zMjcwNjQxMS8zMTYyMjQ5NzQtZmI4MTFmN2UtZjM1Ny00ODIxLWI0YzUtNWU1YTQzMjA5YWI2LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIzVDA3MjUzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMyZGFkY2IzZDRkMzE5Nzk0NmYzNDVkMWY3NGYwNThlY2NhZmU1YzcwYjg1MGI2OWIyYzk5NDZkZTcwOWQ5MmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.YbU22s5i8lyk_h-5gnDJEVXHC_gOCPvGgfM7AY4oA_g) a pending badge like this will look better i think also i should be able to select the pending members.

I thought we couldn't select pending members, will change it.

> ![Screenshot 2024-03-23 at 12 51 05 PM](https://private-user-images.githubusercontent.com/32706411/316225105-c538241e-98c8-4dde-ae79-2c32972fb5d3.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExNzkwMzcsIm5iZiI6MTcxMTE3ODczNywicGF0aCI6Ii8zMjcwNjQxMS8zMTYyMjUxMDUtYzUzODI0MWUtOThjOC00ZGRlLWFlNzktMmMzMjk3MmZiNWQzLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIzVDA3MjUzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc1MGRkZmJlN2FlZjZlOTAxM2M3MGFiODdhMGMyMmI2NGM1OTgyYjkzNjFlODZlMGQyYjg2OTc5MTFiMTI0OTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.BzfT26L3DvlC0ncWVe4-J-M8yXNHH_LMm4Ek9xlror8) should show the pending badge in the selected list as well for pending members.

Will update to show pending badge in the list below. The dropdown is being rendered by react-select so will have to make changes there which would be challenging, will give it a try though.

Amit91848 committed:
fix: show pending badge, open bulk import for multiple emails

Amit91848 commented:
Made changes @SomayChauhan 


https://github.com/calcom/cal.com/assets/74371312/f930e00e-a228-4d8f-ac3c-1f8144f67e0e

Amit91848 commented on packages/features/eventtypes/components/AddMembersWithSwitch.tsx:
```diff
@@ -14,18 +15,21 @@ interface IUserToValue {
   value: `${id || ""}`,
-  label: `${name || email || ""}${!username ? ` (${pendingString})` : ""}`,
-  avatar,
+  // Showing Pending badge if user has not accepted
+  label: `${name || email || ""}${!username && accepted !== false ? ` (${pendingString})` : ""}`,
```
Not sure where this was being used

SomayChauhan replied:
I think changing the common UI component is not a good idea.
we can instead pass the component prop to the <Select/>
take a look how option is implementd

https://github.com/SomayChauhan/cal.com/blob/f3cb6d28a91244c1416b56667517aa6d4e012b8d/apps/web/components/eventtype/EventAvailabilityTab.tsx#L31

Amit91848 commented on packages/ui/components/form/select/Select.tsx:
```diff
@@ -15,6 +17,28 @@ export type SelectProps<
+    <components.Option data={data} {...props}>
+      <>
+        {props.children}
+        {(data as { accepted?: boolean })?.accepted === false && (
```
Kinda hacky, was getting type errors, couldn't figure out how to resolve those.

SomayChauhan replied:
I think changing the common UI component is not a good idea.
we can instead pass the component prop to the <Select/>
take a look how option is implementd

https://github.com/SomayChauhan/cal.com/blob/f3cb6d28a91244c1416b56667517aa6d4e012b8d/apps/web/components/eventtype/EventAvailabilityTab.tsx#L31

Amit91848 replied:
That is what I was trying to find. Made changes tests should pass now.

Amit91848 committed:
Merge branch 'main' into feat/invite_modal_round_robin

SomayChauhan commented:
Great job with the UI @Amit91848, Looking Awesome!!, 
However, Saving the event with pending members as host is throwing an error.
We should allow to add members who are not yet accepted as hosts. But they shouldn't be able to be booked yet.

Amit91848 committed:
fix: error updating event type with pending members

Amit91848 commented:
> Great job with the UI @Amit91848, Looking Awesome!!, However, Saving the event with pending members as host is throwing an error. We should allow to add members who are not yet accepted as hosts. But they shouldn't be able to be booked yet.

Thanks. Fixed saving event. They shouldn't be booked as they don't have any availability yet.

Amit91848 committed:
fix: passing Option proper way

SomayChauhan commented:
Looking Good!!, @Amit91848,
Amazing work

SomayChauhan commented:
@PeerRich should we consider this for collective events as well ?

Amit91848 committed:
Merge branch 'main' into feat/invite_modal_round_robin

Amit91848 committed:
Merge branch 'main' into feat/invite_modal_round_robin

Udit-takkar committed:
Merge branch 'main' into feat/invite_modal_round_robin

joeauyeung commented:
> @PeerRich should we consider this for collective events as well ?

Ideally, we should also be able to do this for collective team events.

CarinaWolli commented on packages/features/eventtypes/components/CheckedTeamSelect.tsx:
```diff
@@ -66,6 +85,11 @@ export const CheckedTeamSelect = ({
               <p className="text-emphasis my-auto ms-3 text-sm">{option.label}</p>
+              {option.accepted === false && (
+                <Badge variant="orange" className="ml-2 text-xs">
+                  Pending
```
translation missing

CarinaWolli commented on packages/features/eventtypes/components/AddMembersWithSwitch.tsx:
```diff
@@ -132,6 +154,19 @@ const AddMembersWithSwitch = ({
+                  e.preventDefault();
+                  setMemberInviteModal(true);
+                  setNoOptionLeft(false);
+                  handleEmailInvite(inputValue);
```
can we remove spaces from the `inputValue`? The input field itself doesn't allow spaces. When testing I typed `free@example, pro@example.com` which gave me the error 'Enter emails'

CarinaWolli reviewed (commented):
Tested and works 👏🏻 @SomayChauhan and @joeauyeung I agree we should also have that for collective events and fixed hosts, should we do it in a follow up or in this PR?

Amit91848 committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat/invite_modal_round_robin

Amit91848 committed:
fix: missing translation, trim emails

Amit91848 commented:
> translation missing

> can we remove spaces from the `inputValue`? The input field itself doesn't allow spaces. When testing I typed `free@example, pro@example.com` which gave me the error 'Enter emails'

Fixed these




> Tested and works 👏🏻 @SomayChauhan and @joeauyeung I agree we should also have that for collective events and fixed hosts, should we do it in a follow up or in this PR?

Won't need much changes, should I include it in this pr itself?

SomayChauhan commented:
> Won't need much changes, should I include it in this pr itself?
Would love to have in this PR itself, Thanks!!

Amit91848 committed:
invite members from fixed hosts and collective events

Amit91848 commented:
> Tested and works 👏🏻 @SomayChauhan and @joeauyeung I agree we should also have that for collective events and fixed hosts, should we do it in a follow up or in this PR?

This works too now


https://github.com/calcom/cal.com/assets/74371312/4c3d6bec-2ed8-4089-b180-fed52a29b6c4

> Also for fixed hosts, we are only showing accepted members, you cannot see the pending member after inviting them, should we show all members and disable the pending ones?

Only this needs to be looked at ig.

PeerRich commented:
awesome PR! asking the obvious here, but inviting people will also add them to the event-type immediately, right?

PeerRich commented:
/tip 100

Amit91848 commented:
> /tip 100

Thanks a lot!! 😍

CarinaWolli commented:
> They shouldn't be booked as they don't have any availability yet.

You can also invite existing users to a team, user that already have an availability. So we still need to fix that pending member can't be booked

CarinaWolli commented:
> Should I make change to add them to event type? Since you cannot have pending users as fixed hosts, user would be added to event type only if they are invited from round robin hosts.

Why can't we have pending users as fixed hosts?

Amit91848 commented:
> > Should I make change to add them to event type? Since you cannot have pending users as fixed hosts, user would be added to event type only if they are invited from round robin hosts.
> 
> Why can't we have pending users as fixed hosts?

https://github.com/calcom/cal.com/blob/93bb37702c8790e4b56b74a04e6bcd137ca9d807/apps/web/components/eventtype/EventTeamTab.tsx#L384-L388

https://github.com/calcom/cal.com/blob/93bb37702c8790e4b56b74a04e6bcd137ca9d807/apps/web/components/eventtype/EventTeamTab.tsx#L432-L436

Pending members are being filtered before being passed to Fixed Hosts so I assumed that was the case, is it not?

Amit91848 commented:
> You can also invite existing users to a team, user that already have an availability. So we still need to fix that pending member can't be booked.

Yeah my bad, will look into it.

Amit91848 commented:
> Ah, I see. I assume this was needed because pending members often have no availability and therefore would block the whole availability then. But if we fix that pending members can't be booked yet and won't be included in the availabilities, then this shouldn't be an issue

Okay makes sense 👍

Amit91848 committed:
filter pending users to avoid blocking availability and allows booking

Amit91848 commented:
https://github.com/calcom/cal.com/assets/74371312/3505a2f6-97ba-4919-9e4e-3d85a0eb9143


Does not check for pending users availability and also can book with pending hosts.

CarinaWolli commented on packages/trpc/server/routers/viewer/eventTypes/update.handler.ts:
```diff
@@ -225,11 +226,15 @@ export const updateHandler = async ({ ctx, input }: UpdateOptions) => {
         where: {
           teamId,
-          accepted: true,
+          accepted: !eventIsOrWillBeRoundRobin ? true : undefined,
```
adding pending member to collective event type throws FORBIDDEN

CarinaWolli reviewed (changes requested):
Already assigned team members are not disabled anymore and I can add them to both dropdowns, saving it throws and error:
![Screenshot 2024-05-02 at 12 40 36 PM](https://github.com/calcom/cal.com/assets/30310907/b525786f-a1a9-4b41-9c96-0f9d29d163c3)

Amit91848 committed:
fix: disable selected members

Amit91848 committed:
fix: error thrown while saving collective

Amit91848 committed:
code refactor

CarinaWolli committed:
Merge branch 'main' into feat/invite_modal_round_robin

CarinaWolli commented:
@Amit91848 let me know when this is ready for review again 🙏

Amit91848 commented:
> @Amit91848 let me know when this is ready for review again 🙏

Will do, all work is done. Few tests are breaking looking into those.

Amit91848 committed:
Merge branch 'main' into feat/invite_modal_round_robin

Amit91848 committed:
fix: unit tests

Amit91848 commented:
Some are still failing but I don't think they are related to my changes.

CarinaWolli commented:
> awesome PR! asking the obvious here, but inviting people will also add them to the event-type immediately, right?

@Amit91848 can we do that? Other than that everything looked good 🙏

Amit91848 commented:
Not sure where it would fit tho. 

Maybe we can pass eventTypeId and isFixed to createInvite and update eventType there?
https://github.com/calcom/cal.com/blob/63dd3576646ee591d7b71c75a4d547ebb261b73e/packages/trpc/server/routers/viewer/teams/createInvite.handler.ts#L15-L19

Or we can call updateHandler mutation onSuccess of createInvite mutation
https://github.com/calcom/cal.com/blob/ed5463fb33a139bbb926a4950050026463f8e7c7/apps/web/components/eventtype/EventTeamTab.tsx#L419-L440

Lmk what you think

Amit91848 commented:
> 
> Or we can call updateHandler mutation onSuccess of createInvite mutation
> 
> https://github.com/calcom/cal.com/blob/ed5463fb33a139bbb926a4950050026463f8e7c7/apps/web/components/eventtype/EventTeamTab.tsx#L419-L440
> 
> Lmk what you think

I was trying this, createInviteMembers returns only email of users added not the id, thought about comparing it with teamMembers to get the id but teamMembers is stuck in old state until the execution of onSuccess is completed and component re renders. 

> awesome PR! asking the obvious here, but inviting people will also add them to the event-type immediately, right?


Not sure how to achieve this.

Amit91848 committed:
Merge branch 'main' into feat/invite_modal_round_robin

CarinaWolli commented:
> I was trying this, createInviteMembers returns only email of users added not the id, thought about comparing it with teamMembers to get the id but teamMembers is stuck in old state until the execution of onSuccess is completed and component re renders.

What about newly fetching the event type after `eventTypes.get.invalidate`. Fetching it like that should work: 
`const newEventType = await utils.viewer.eventTypes.get.fetch({ id: getValues("id") });
`

Amit91848 committed:
add invited members to event type

Amit91848 committed:
code refactor

Amit91848 commented:
https://github.com/calcom/cal.com/assets/74371312/86053e6f-e245-4cec-89d0-60cb044d5ea8

Made changes but first invite doesn't change form state so event type isn't being updated. isFormDirty is false for some reason couldn't figure out why

https://github.com/calcom/cal.com/blob/2c8c8368a01f65f5f18a04b95530c0a6633bfbeb/apps/web/modules/event-types/views/event-types-single-view.tsx#L502-L505

CarinaWolli commented:
> Made changes but first invite doesn't change form state so event type isn't being updated. 

I justed tested your current version and the first invite seems to work as it should. They get invited, then added as hosts and the save button is enabled if you decide to save it. However, when adding a second host it saves automatically for me. Am I misunderstanding something?

Amit91848 committed:
fix: update only form state

Amit91848 committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat/invite_modal_round_robin

Amit91848 committed:
fix: unit test

Amit91848 commented:
Made changes @CarinaWolli please review

CarinaWolli reviewed (dismissed):
Works great now 🙌🏻 Thank you for your contribution 🙏

PeerRich commented:
flaky tests: https://github.com/calcom/cal.com/actions/runs/9421468407/job/25955940198

Amit91848 commented:
> flaky tests: https://github.com/calcom/cal.com/actions/runs/9421468407/job/25955940198

https://github.com/calcom/cal.com/blob/cd888be61e57b77203576511c5f73c836f6d5013/apps/web/playwright/dynamic-booking-pages.e2e.ts#L131-L133

This page is throwing a 404, couldn't figure out why

PeerRich committed:
Merge branch 'main' into feat/invite_modal_round_robin

ibex088 reviewed (changes requested):
Hi @Amit91848 
you need to set `NEXT_PUBLIC_IS_E2E=1` 
to run org tests locally

also wanted to add that this test is passing on main and failing on this branch.
can confirm that the logic you added on `packages/trpc/server/routers/viewer/slots/util.ts` is causing the tests to fail

zomars commented on packages/trpc/server/routers/viewer/slots/util.ts:
```diff
@@ -448,6 +458,14 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Pro
   let usersWithCredentials = hosts.map(({ isFixed, user }) => ({ isFixed, ...user }));
 
+  if (eventType.schedulingType) {
+    //remove pending members  to avoid blocking availability for all
+    const teamMembers = eventType.team?.members || [];
+    usersWithCredentials = usersWithCredentials.filter((user) =>
+      teamMembers.some((member) => member.userId === user.id && member.accepted === true)
+    );
```
Can we do this at DB query level?

zomars commented on packages/trpc/server/routers/viewer/eventTypes/update.handler.ts:
```diff
@@ -225,12 +226,10 @@ export const updateHandler = async ({ ctx, input }: UpdateOptions) => {
       (await ctx.prisma.membership.findMany({
         where: {
           teamId,
-          accepted: true,
```
Why was this removed?

Amit91848 replied:
https://github.com/calcom/cal.com/blob/bef1bf6afeedd270c8e8b154af5621ed3abb7f63/packages/trpc/server/routers/viewer/eventTypes/update.handler.ts#L235-L243

Earlier pending members couldn't be assigned as hosts so we only needed to check for accepted ones, this isn't the case now

zomars reviewed (changes requested):
Thank you for your contribution 🙏 

Self-review seems to be missing. Or I couldn't find it.

Can we have unit/integration tests to prove this feature works and it won't break with future changes?

My main concerns are:

- Does pending members affect the RR availability in any ways? If yes/no, can we prove it with a test?
- What happens if the added members rejects the invite? Should it be removed from the RR list? Can we make a test for that?

Also I have concerns about code quality:

- I'm noticing quite some prop-drilling
- A lot of repeated code, mostly on team selects . Instead of extracting to a reusable variable. This is prone to de syncing issues.

I would love to see this feature properly tested in order to increase confidence to merge to main. RN I don't feel confident merging in the current state.

CarinaWolli committed:
fix dynamic event type availability

CarinaWolli commented on packages/trpc/server/routers/viewer/slots/util.ts:
```diff
@@ -448,6 +458,14 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Pro
 
   let usersWithCredentials = hosts.map(({ isFixed, user }) => ({ isFixed, ...user }));
 
+  if (eventType.schedulingType && eventType.team) {
```
We need to check here if it is a team even-type because dynamic event types are Collective event types but don't have team members. This resulted in empty availability for dynamic event types and caused `dynamic-bookings-pages.e2e.ts` to fail

Amit91848 committed:
remove from host on rejecting invitation

Amit91848 commented:
> Can we have unit/integration tests to prove this feature works and it won't break with future changes?

Will add it.

> My main concerns are:
> 
> * Does pending members affect the RR availability in any ways? If yes/no, can we prove it with a test?

Pending members are filtered out while calculating slots, so only non pending hosts are taken into consideration. If all hosts are pending then no availability is shown.

> * What happens if the added members rejects the invite? Should it be removed from the RR list? Can we make a test for that?

I assumed their assignment as hosts were also being deleted. Just checked and this isn't the case. Added changes for that

> Also I have concerns about code quality:
> 
> * I'm noticing quite some prop-drilling
> * A lot of repeated code, mostly on team selects . Instead of extracting to a reusable variable. This is prone to de syncing issues.
> 
> I would love to see this feature properly tested in order to increase confidence to merge to main. RN I don't feel confident merging in the current state.

Sure thing, working on these

Amit91848 committed:
use db query to filter

Amit91848 committed:
chore: booking rr e2e

keithwillcode commented:
This PR is being closed due to a lack of activity. Please reopen if work is to continue.
